packages/google_maps_flutter/google_maps_flutter/example/lib/clustering.dart

```diff
@@ -0,0 +1,237 @@
+  }
+}
+
+class ClusteringBody extends StatefulWidget {
+  const ClusteringBody({Key? key}) : super(key: key);
+
+  @override
+  State<StatefulWidget> createState() => ClusteringBodyState();
+}
+
+typedef MarkerUpdateAction = Marker Function(Marker marker);
+
+class ClusteringBodyState extends State<ClusteringBody> {
+  ClusteringBodyState();
+  static const LatLng center = LatLng(-33.86711, 151.1547171);
+  static const double _scaleFactor = 0.05;
+
+  GoogleMapController? controller;
+  Map<ClusterManagerId, ClusterManager> clusterManagers =
+      <ClusterManagerId, ClusterManager>{};
+  Map<MarkerId, Marker> markers = <MarkerId, Marker>{};
+  MarkerId? selectedMarker;
+  int _clusterManagerIdCounter = 1;
+  int _markerIdCounter = 1;
+  Cluster? lastCluster;
+
+  // ignore: use_setters_to_change_properties
+  void _onMapCreated(GoogleMapController controller) {
+    this.controller = controller;
+  }
+
+  @override
+  void dispose() {
+    super.dispose();
+  }
+
+  void _onMarkerTapped(MarkerId markerId) {
+    final Marker? tappedMarker = markers[markerId];
+    if (tappedMarker != null) {
+      setState(() {
+        final MarkerId? previousMarkerId = selectedMarker;
+        if (previousMarkerId != null && markers.containsKey(previousMarkerId)) {
+          final Marker resetOld = markers[previousMarkerId]!
+              .copyWith(iconParam: BitmapDescriptor.defaultMarker);
+          markers[previousMarkerId] = resetOld;
+        }
+        selectedMarker = markerId;
+        final Marker newMarker = tappedMarker.copyWith(
+          iconParam: BitmapDescriptor.defaultMarkerWithHue(
+            BitmapDescriptor.hueGreen,
+          ),
+        );
+        markers[markerId] = newMarker;
+      });
+    }
+  }
+
+  void _addClusterManager() {
+    final int clusterManagerCount = clusterManagers.length;
+
+    if (clusterManagerCount == 3) {
+      return;
+    }
+
+    final String clusterManagerIdVal =
+        'cluster_manager_id_$_clusterManagerIdCounter';
+    _clusterManagerIdCounter++;
+    final ClusterManagerId clusterManagerId =
+        ClusterManagerId(clusterManagerIdVal);
+
+    final ClusterManager clusterManager = ClusterManager(
+      clusterManagerId: clusterManagerId,
+      onClusterTap: (Cluster cluster) => setState(() {
+        lastCluster = cluster;
+      }),
+    );
+
+    setState(() {
+      clusterManagers[clusterManagerId] = clusterManager;
+    });
+    _addMarkersToCluster(clusterManager);
+  }
+
+  void _removeClusterManager(ClusterManager clusterManager) {
+    setState(() {
+      // Remove markers managed by cluster manager to be removed.
+      markers.removeWhere((MarkerId key, Marker marker) =>
+          marker.clusterManagerId == clusterManager.clusterManagerId);
+      // Remove cluster manager.
+      clusterManagers.remove(clusterManager.clusterManagerId);
+    });
+  }
+
+  void _addMarkersToCluster(ClusterManager clusterManager) {
+    for (int i = 0; i < 15; i++) {
```

Comment:
This just adds 15 markers to cluster at once, this could be constant as well, but keeping examples as compact as possible is also important. We'll add comment before this line to explain the functionality